packages/core/src/verify.ts

```diff
@@ -0,0 +1,166 @@
+): AsyncGenerator<VerifyResult> {
+  const deploymentLoader = new FileDeploymentLoader(deploymentDir);
+
+  const deploymentState = await loadDeploymentState(deploymentLoader);
+
+  if (deploymentState === undefined) {
+    throw new IgnitionError(ERRORS.VERIFY.UNINITIALIZED_DEPLOYMENT, {
+      deploymentDir,
+    });
+  }
+
+  const chainConfig = resolveChainConfig(deploymentState, customChains);
+
+  const deploymentExStates = findExecutionStatesByType(
+    ExecutionSateType.DEPLOYMENT_EXECUTION_STATE,
+    deploymentState
+  );
+
+  if (deploymentExStates.length === 0) {
+    throw new IgnitionError(ERRORS.VERIFY.NO_CONTRACTS_DEPLOYED, {
+      deploymentDir,
+    });
+  }
+
+  for (const exState of deploymentExStates) {
+    const verifyInfo = await convertExStateToVerifyInfo(
+      exState,
+      deploymentLoader
+    );
+
+    const verifyResult: VerifyResult = [chainConfig, verifyInfo];
+
+    yield verifyResult;
+  }
+}
+
+function resolveChainConfig(
+  deploymentState: DeploymentState,
+  customChains: ChainConfig[]
+) {
+  // implementation note:
+  // if a user has set a custom chain with the same chainId as a builtin chain,
+  // the custom chain will be used instead of the builtin chain
+  const chainConfig = [...customChains, ...builtinChains].find(
+    (c) => c.chainId === deploymentState.chainId
+  );
+
+  if (chainConfig === undefined) {
+    throw new IgnitionError(ERRORS.VERIFY.UNSUPPORTED_CHAIN, {
+      chainId: deploymentState.chainId,
+    });
+  }
+
+  return chainConfig;
+}
+
+async function convertExStateToVerifyInfo(
+  exState: DeploymentExecutionState,
+  deploymentLoader: FileDeploymentLoader
+) {
+  const [buildInfo, artifact] = await Promise.all([
+    deploymentLoader.readBuildInfo(exState.artifactId),
+    deploymentLoader.loadArtifact(exState.artifactId),
+  ]);
+
+  const { contractName, constructorArgs, libraries } = exState;
+
+  assertIgnitionInvariant(
+    exState.result !== undefined &&
+      exState.result.type === ExecutionResultType.SUCCESS,
+    `Deployment execution state ${exState.id} should have a successful result to retrieve address`
+  );
+
+  const verifyInfo = {
+    address: exState.result.address,
+    compilerVersion: buildInfo.solcLongVersion.startsWith("v")
+      ? buildInfo.solcLongVersion
+      : `v${buildInfo.solcLongVersion}`,
+    sourceCode: prepareInputBasedOn(buildInfo, artifact, libraries),
+    name: `${artifact.sourceName}:${contractName}`,
+    args: encodeDeploymentArguments(artifact, constructorArgs),
+  };
+
+  return verifyInfo;
+}
+
+function prepareInputBasedOn(
+  buildInfo: BuildInfo,
+  artifact: Artifact,
+  libraries: Record<string, string>
+): string {
+  const sourceToLibraryAddresses = resolveLibraryInfoForArtifact(
+    artifact,
+    libraries
+  );
+
+  if (sourceToLibraryAddresses === null) {
+    return JSON.stringify(buildInfo.input);
+  }
+
+  const { input } = buildInfo;
+  input.settings.libraries = sourceToLibraryAddresses;
```

Comment:
This will lead to a different metadata hash. If we can avoid this, we should. Can we pass the libraries to the `Etherscan` object as a separate param?

Reply:
Unfortunately the `Etherscan` class doesn't accept libraries as a param, however, `hardhat-verify` seems to be adding library info [the same way](https://github.com/NomicFoundation/hardhat/blob/main/packages/hardhat-verify/src/internal/tasks/etherscan.ts#L298).